2018/003-offline-chart-repository.md

```diff
@@ -0,0 +1,441 @@
+
+Updating the chart image can be done in a similar way to how the SUSE CaaS
+Platform teams implements their automated packaging pipeline:
+
+  * Multiple concourse jobs are defined for each chart package
+  * Each concourse job monitors the changes submitted to the master branch of
+    the git repository containing the source code of the chart
+  * The OBS package is automatically updated by concourse
+
+This will automatically trigger a rebuild of the RPM package of the chart and
+later of the chart container image.
+
+**Note well:** it would also be possible to build helm charts based on openSUSE inside of the OBS instance and then have them synchronized inside of IBS to build their SLE counterparts. This is similar to what the SUSE CaaS team is doing with other bits and pieces of Kubic.
+
+#### Online helm chart repository
+
+The SUSE Customer Center team is currently maintaining the hosted version of the SUSE chart repository.
+
+This is currently implemented by exposing the contents of a S3 bucket. The S3 bucket contains all the `.tgz` files and the `index.yaml` that make the repository. The helm chart creation, the index generation and the upload to S3 is currently automated via a concourse pipeline described [here](https://github.com/SUSE/cloudfoundry/wiki/Building-SUSE-Cloud-Application-Platform#publishing-to-the-helm-repository-httpskubernetes-chartssusecom).
+
+If this proposal is accepted the existing automation pipeline could be extended to:
+
+  * Download the latest version of the SUSE helm chart image: `skopeo copy docker://registry.suse.com/suse/kubernetes-charts:latest`. oci:///tmp/automation/suse-kubernetes-charts:latest`.
+  * Extract the contents of the image: `umoci unpack --rootless --image /tmp/suse-kubernetes-charts:latest /tmp/automation/suse-kubernetes-charts-unpacked`.
+  * Copy the helm charts shipped with the image: `mv /tmp/automation/suse-kubernetes-charts-unpacked/rootfs/usr/share/suse-kubernetes-charts/ /tmp/automation/`.
+  * Recreate the `index.yaml` file: `helm repo index /tmp/automation/suse-kubernetes-charts`.
+  * Upload all the contents of `/tmp/automation/suse-kubernetes-charts` to S3.
+
+Note well: neither `skopeo` nor `umoci` require access to a docker daemon, hence they can be run inside of a container in an easy fashion.
+
+It would also be possible to not use the S3 bucket at all and just run an instance of the latest version of the `registry.suse.com/suse/kubernetes-charts` container image.
+
+
+
+### Dependencies
+
+> Highlight how the change may affect the rest of the product (new components,
+> modifications in other areas), or other teams/products.
+
+The adoption of this workflow would have an impact on the current online helm chart repository. This has been outlined in the previous section (see *“Online helm chart repository”*).
+
+### Concerns and Unresolved Questions
+
+> List any concerns, unknowns, and generally unresolved questions etc.
+
```

Comment:
When we get into the implementation of such things, we always struggle with the tagging of the images. For example, helm charts may contain "image: foo:abcd1234" but then the image gets rebuild and the tag does not work anymore. I think we should consider this case. This is a general issue with the images and we already have this problem, so this document may not be the place for the solution, but I think is worth mentioning it that this needs to be addressed.

Reply:
+1